plutus-playground-client/src/Chain/Types.purs

```diff
@@ -1,42 +1,101 @@
+  = FocusTx TxId
 
-_FocusTx :: Iso' ChainFocus AnnotatedTx
+_FocusTx :: Iso' ChainFocus TxId
 _FocusTx = iso get set
   where
-    get (FocusTx annotatedTx) = annotatedTx
-    set = FocusTx
+  get (FocusTx txId) = txId
+
+  set = FocusTx
 
 derive instance genericChainFocus :: Generic ChainFocus _
 
 instance showChainFocus :: Show ChainFocus where
   show = genericShow
 
-type State =
-  { chainFocus :: Maybe ChainFocus
-  , chainFocusAppearing :: Boolean
-  , chainFocusAge :: Ordering
-  }
+type State
+  = { chainFocus :: Maybe ChainFocus
+    , chainFocusAppearing :: Boolean
+    , chainFocusAge :: Ordering
+    }
+
+_chainFocus :: forall r a. Lens' { chainFocus :: a | r } a
+_chainFocus = prop (SProxy :: SProxy "chainFocus")
+
+_chainFocusAppearing :: forall r a. Lens' { chainFocusAppearing :: a | r } a
+_chainFocusAppearing = prop (SProxy :: SProxy "chainFocusAppearing")
+
+_chainFocusAge :: forall r a. Lens' { chainFocusAge :: a | r } a
+_chainFocusAge = prop (SProxy :: SProxy "chainFocusAge")
 
 _sequenceId :: Lens' AnnotatedTx SequenceId
 _sequenceId = _Newtype <<< prop (SProxy :: SProxy "sequenceId")
 
+_dereferencedInputs :: Lens' AnnotatedTx (Array DereferencedInput)
+_dereferencedInputs = _Newtype <<< prop (SProxy :: SProxy "dereferencedInputs")
+
+_originalInput :: Lens' DereferencedInput (TxInOf String)
+_originalInput = _Newtype <<< prop (SProxy :: SProxy "originalInput")
+
+_txIdOf :: Lens' AnnotatedTx TxId
+_txIdOf = _Newtype <<< prop (SProxy :: SProxy "txId")
+
+_tx :: Lens' AnnotatedTx Tx
+_tx = _Newtype <<< prop (SProxy :: SProxy "tx")
+
+_txInputs :: Lens' Tx (Array (TxInOf String))
+_txInputs = _Newtype <<< prop (SProxy :: SProxy "txInputs")
+
+_txOutputs :: Lens' Tx (Array (TxOutOf String))
+_txOutputs = _Newtype <<< prop (SProxy :: SProxy "txOutputs")
+
+_txId :: forall a. Lens' (TxIdOf a) a
+_txId = _Newtype <<< prop (SProxy :: SProxy "getTxId")
+
+_txInRef :: forall a. Lens' (TxInOf a) (TxOutRefOf a)
+_txInRef = _Newtype <<< prop (SProxy :: SProxy "txInRef")
+
+_txOutRefId :: forall a. Lens' (TxOutRefOf a) (TxIdOf a)
+_txOutRefId = _Newtype <<< prop (SProxy :: SProxy "txOutRefId")
+
 toBeneficialOwner :: TxOutOf String -> BeneficialOwner
-toBeneficialOwner (TxOutOf {txOutType, txOutAddress}) =
-    case txOutType of
-        PayToPubKey pubKey -> OwnedByPubKey pubKey
-        PayToScript _      -> OwnedByScript txOutAddress
+toBeneficialOwner (TxOutOf { txOutType, txOutAddress }) = case txOutType of
+  PayToPubKey pubKey -> OwnedByPubKey pubKey
+  PayToScript _ -> OwnedByScript txOutAddress
+
+_findTx :: forall m. Monoid m => TxId -> Fold' m (Array (Array AnnotatedTx)) AnnotatedTx
+_findTx focussedTxId = (traversed <<< traversed <<< filtered isAnnotationOf)
+  where
+  isAnnotationOf :: AnnotatedTx -> Boolean
+  isAnnotationOf (AnnotatedTx { txId }) = txId == focussedTxId
+
+-- | Where is this output consumed?
+findConsumptionPoint :: Int -> TxIdOf String -> Array (Array AnnotatedTx) -> Maybe AnnotatedTx
```

Comment:
Not really. The blockchain is an array of arrays, and I'm preserving that structure. Arguably we should be newtyping it though... 🤔

Reply:
The main reason I said this is because it's hard to understand, usually an array of arrays implies some calculation happened that needs flattening at the end rather than an actual designed structure so I think newtypes would fit in well here

Reply:
for the future though I guess